### ISSUE TITLE
BLE001 should be allowed if function returns NoReturn

### ISSUE THREAD

**Issue by rodrigoscc** [astral-sh/ruff#23150]
### Summary

Hello!

This is a question regarding the BLE001 rule. 

The rule is useful but I often have issues with false positives when handling the exception inside the catch block. I understand that detecting every case in this specific rule is hard. However, to detect cases where an exception is re-raised or wrapped in another exception within a function, ruff could make use of NoReturn. So, should ruff allow catching blind exceptions when a function called inside the catch block returns NoReturn?

```python
from typing import NoReturn

def just_raise(exc: Exception) -> NoReturn:
    raise Exception from exc

def main():
    try:
        raise Exception
    except Exception as exc: # This line reports BLE001 error
        just_raise(exc)
```

Thanks for your time!

**Comment by ntBre** [comment]
Thanks for the suggestion! This sounds reasonable to me, but I'm not sure how reliably we'll be able to detect this case without type inference or at least multi-file analysis.

**Comment by amyreese** [comment]
I'm not sure I understand the use case here, or why the lint rule should treat it differently just because the function called in the exception handler has a `NoReturn` signature. It's still catching `Exception` rather than a more specific type.

**Comment by ntBre** [comment]
The rule has several exceptions ([docs](https://docs.astral.sh/ruff/rules/blind-except/#example)):

> Exceptions that are re-raised will not be flagged, as they're expected to be caught elsewhere:

> Exceptions that are logged with exc_info enabled will not be flagged, as this is a common pattern for propagating exception traces:

If you assume that `NoReturn` means the function is going to re-raise the exception, then I think this makes sense as another modification to the rule that indicates the exception is being handled elsewhere. Is that not a common use for `NoReturn`? The [typing docs](https://docs.python.org/3/library/typing.html#typing.NoReturn) just say:

> They can be used to indicate that a function never returns, such as [sys.exit()](https://docs.python.org/3/library/sys.html#sys.exit):

**Comment by amyreese** [comment]
Passing an exception to a `NoReturn` expecting that function to re-raise the same exception is not a common pattern IME.  I'd be more inclined to focus on the *parameter* type taking `Exception` rather than whether or not it has `NoReturn`, but I still feel like `BLE001` in either case is valid, again because it's pointing out that you're handling an overly-broad exception type, regardless of what you might do with it.  If it's an intentional choice, then a suppression comment is the correct solution.

**Comment by ntBre** [comment]
> it's pointing out that you're handling an overly-broad exception type, regardless of what you might do with it. If it's an intentional choice, then a suppression comment is the correct solution.

This could apply to the existing exceptions as well. The third ecosystem case I checked for this rule looks like an exact instance of this: https://github.com/prefecthq/prefect/blob/main/src/prefect/cli/block.py#L487

**Comment by Avasam** [comment]
Just my 2 cents, you want to avoid silencing an overly broad exception right ? (hence why logging or reraising is a valid exception to this rule). To me it's likely that a method that takes in an exception and never returns is essentially reraising (maybe with extra info, maybe by further refining the exception, etc).
Type-checking would take care of making sure you're not passing an overly broad exception to `just_raise` if that's not the intent.
So I can see the argument for allowing this pattern.

The question is still whether this is even worth trying to implement without type inference.

**Comment by danparizher** [comment]
To @amyreese's point about it still being a broad catch: while that's true, the "danger" of a broad catch is significantly mitigated if the control flow is guaranteed to stop or propagate the error. For users who intentionally wrap their exception handling in a terminating helper, this would remove a persistent false positive.

Technically, I think we can actually solve this without needing full-blown cross-file type inference. There is already an `is_noreturn_func` helper used in the `RET` (flake8-return) rules, which could be reused here.

What if we centralize the detection to a shared utility and go from there? I'd be glad to give it a shot if it makes sense.

**Comment by a1tus** [comment]
Sorry if it's a little bit off-topic but I suppose it's related enough.

We have similar problem in our project. Our common pattern is:

```python
from tracing.logs import log_exc
try:
  1 / 0
except Exception:  # here we have to use "noqa: BLE001"
  log_exc('...')
```

And https://docs.astral.sh/ruff/settings/#lint_logger-objects does not help, because it requires some object, on which `.exception` method would be called.

May be something like `logger-functions` can be added that will simply check names being called inside of except block?
Or even logger-objects can be treated more widely (allow both `.exception` and direct calls).

**Comment by ntBre** [comment]
@a1tus I think that request is tracked in https://github.com/astral-sh/ruff/issues/12613.

**Comment by a1tus** [comment]
Just for info, I've created PR https://github.com/astral-sh/ruff/pull/23783
